src/category_theory/limits/shapes/zero.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2019 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.limits.shapes.terminal
+
+/-!
+# Zero morphisms and zero objects
+
+A category "has zero morphisms" if there is a designated "zero morphism" in each morphism space,
+and compositions of zero morphisms with anything give the zero morphism. (Notice this is extra
+structure, not merely a property.)
+
+A category "has a zero object" if it has an object which is both initial and terminal. Having a
+zero object provides zero morphisms, as the unique morphisms factoring through the zero object.
+
+## References
+* https://en.wikipedia.org/wiki/Zero_morphism
+-/
+
+universes v u
+
+open category_theory
+
+namespace category_theory.limits
+
+variables (C : Type u) [𝒞 : category.{v} C]
+include 𝒞
+
+/-- A category "has zero morphisms" if there is a designated "zero morphism" in each morphism space,
+and compositions of zero morphisms with anything give the zero morphism. -/
+class has_zero_morphisms :=
```

Comment:
Shall we provide `has_one_morphisms` for multiplicative group objects? (E.g. in the category `Group` it would seem weird to denote the constant hom with `0`. Also, it's "zero object" would more reasonably be denoted with `1`.

Reply:
I'm dubious. "Having a zero object" is standard terminology, but no one is ever going to say "having a one object" in the real world. It is nice to be able to write `1` for the trivial group in a short exact sequence, I guess. Talking about `kernel` and `image` won't require knowing whether the zero morphism is called `0` or `1`, in any case.

Would you actually write `1 : G \hom H` for the morphism that sends everything to the identity in `H`? That feels a bit weird to me.

Reply:
Yeah, I agree that it is a very special case. Let's just ignore it.